pkg/proxy/faultinjection/manager.go

```diff
@@ -0,0 +1,593 @@
+		indexForResource(namespace, apiGroup, "*", "*"),
+		// three wildcard
+		indexForResource("*", "*", "*", verb),
+		indexForResource("*", "*", resource, "*"),
+		indexForResource("*", apiGroup, "*", "*"),
+		indexForResource(namespace, "*", "*", "*"),
+		// four wildcard
+		indexForResource("*", "*", "*", "*"),
+	}
+	return result
+}
+
+func withFaultInjection(validator Validator, handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			// if this happens, the handler chain isn't setup correctly because there is no request info
+			responsewriters.InternalError(w, req, errors.New("no RequestInfo found in the context"))
+			return
+		}
+		result := validator.ValidateResource(requestInfo.Namespace, requestInfo.APIGroup, requestInfo.Resource, requestInfo.Verb)
+
+		if !result.Allowed {
+			apiErr := httpToAPIError(int(result.ErrCode), result.Message)
+			if apiErr.Code != http.StatusOK {
+				w.Header().Set("Content-Type", "application/json")
+				w.WriteHeader(int(apiErr.Code))
+				json.NewEncoder(w).Encode(apiErr)
+				logger.Info("faultinjection rule", fmt.Sprintf("fault injection, %s, %s,%d", result.Reason, result.Message, result.ErrCode))
+				return
+			}
+		}
+
+		realEndPointUrl, err := url.Parse(req.Header.Get("Mesh-Real-Endpoint"))
+		logger.Info("receive proxy-host:%s, proxy-method:%s, Mesh-Real-Endpoint:%s", realEndPointUrl.Host, req.Method, req.Header.Get("Mesh-Real-Endpoint"))
+		if err != nil || realEndPointUrl == nil {
+			logger.Error(err, "Request Header Mesh-Real-Endpoint Parse Error")
+			http.Error(w, fmt.Sprintf("Can not find real endpoint in header %s", err), http.StatusInternalServerError)
+			return
+		}
+		// if enableRestFaultInjection {
+		// 	result := pkgfi.ValidateRest(realEndPointUrl.Host, req.Method)
+		// 	if result.Allowed == false {
+		// 		logger.Error("ErrorTProxy: %s %s  ValidateTrafficIntercept NOPASSED ,checkresult:\t%s", realEndPointUrl.Host, req.Method, result.Reason)
+		// 		http.Error(w, fmt.Sprintf("Forbidden by  ValidateTrafficIntercept breaker, %s, %s", result.Message, result.Reason), http.StatusForbidden)
+		// 		return
+		// 	}
+		// }
+
+		// ValidateRest check
+		logger.Info("start ValidateRest checkrule %s %s", realEndPointUrl.Host, req.Method)
+		validateresult := validator.ValidateRest(req.Header.Get("Mesh-Real-Endpoint"), req.Method)
+		if !validateresult.Allowed {
+			logger.Info("ErrorTProxy: %s %s  ValidateRest NOPASSED ,checkresult:%t, validateresultReason:%s", req.Header.Get("Mesh-Real-Endpoint"), req.Method, validateresult.Allowed, validateresult.Reason)
+			// http.Error(w, fmt.Sprintf("Forbidden by circuit ValidateRest breaker, %s, %s", validateresult.Message, validateresult.Reason), http.StatusForbidden)
+			// return
+		}
+		logger.Info("TProxy: %s %s ValidateRest check PASSED", realEndPointUrl.Host, req.Method)
+
```

Comment:
`withFaultInjection` only takes effect for apiserver requests.
donot validate rest here

Reply:
add http proxy in main.go